### DISCUSSION TITLE
Documentation/tooling request: building custom images

### DISCUSSION THREAD

**Discussion by MrDrMcCoy** [MichaIng/DietPi#6015] [Q&A]
I love what DietPi brings to the table, but I have some obscure needs that aren't met by the default images. Rather than request that the team build special images just for me, I would love to just do it myself, as I currently am able to [with Armbian](https://docs.armbian.com/Developer-Guide_Build-Preparation/).

Currently, the things that I (and doubtlessly others) need from a custom build system are as follows:

1. Custom kernel config
2. Btrfs rootfs
3. Adding arbitrary files and configs
4. Base off Debian Testing instead of just Stable

Being able to do this with DietPi instead of Armbian would require less hacks to de-cruft and doubtlessly end up being more stable. As of now, I have not been able to find any info on how DietPi is currently built. Even a dump of whatever is used now would be helpful, and I would be happy to contribute back anything I end up building as far as scripts go.

**Comment by Joulinar** [comment]
Our base image depends on SBC as we use different types of images like Armbian, plain Debian or Raspberry OS. Furthermore, we don't do own kernel development. Most of the time, we use stock kernel / mainline kernel provided by the base image. We already offer Debian Bookworm images. Feel free to use/test them https://dietpi.com/downloads/images/

**Comment by MichaIng** [comment]
To be precise: All our images are build from scratch via `debootstrap` in the meantime. But indeed we do not compile kernel packages ourselves but use the repos of Armbian, Debian and Rasperry Pi OS respectively for kernel and bootloader.

The [dietpi-installer](https://dietpi.com/docs/hardware/#make-your-own-distribution) can be used to turn any Debian image into a DietPi image. For our own images, we use [dietpi-build](https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-build) to generate the base image. It also allows you to choose between Debian Bullseye and Bookworm as well as choose the root filesystem type. But the letter, I think, still needs some updates to use Btrfs and F2FS, as I haven't tested it since a while. That script also includes minimising and archiving the image, which can otherwise be done separately with [dietpi-imager](https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-imager).
